### PR TITLE
Add threaded callbacks for functions of times

### PR DESCRIPTION
**Pull request by nilsdeppe**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/Callback.hpp [hunk 1]
```diff
@@ -97,6 +97,71 @@ class SimpleActionCallback<SimpleAction, Proxy> : public Callback {
   std::decay_t<Proxy> proxy_{};
 };
 
+/// Wraps a call to a threaded action and its arguments.
+/// Can be invoked only once.
+template <typename ThreadedAction, typename Proxy, typename... Args>
+class ThreadedActionCallback : public Callback {
```
Comment (knelli2) [inline src/Parallel/Callback.hpp:103]: We have a test for these in `Unit/Parallel/Test_Callback.cpp`. Could you add some there? I know it's a bit more work, but just so we are guaranteed that they are working

File: src/ParallelAlgorithms/Actions/FunctionsOfTimeAreReady.hpp [hunk 2]
```diff
@@ -143,6 +143,32 @@ bool functions_of_time_are_ready_simple_action_callback(
...
 
+/// \ingroup ComputationalDomainGroup
+/// Check that functions of time are up-to-date.
+///
+/// Check that functions of time in \p CacheTag with names in \p
+/// functions_to_check are ready at time \p time.  If  \p functions_to_check is
+/// a `std::nullopt`, checks all functions in \p CacheTag.  If any function is
+/// not ready, schedules a `Parallel::ThreadedActionCallback` with the
+/// GlobalCache which calls the threaded action passed in as a template
+/// parameter. The `Args` are moved into the callback.
```
Comment (knelli2) [inline src/ParallelAlgorithms/Actions/FunctionsOfTimeAreReady.hpp]: They're forwarded, not moved. There's a difference, right?

**Review by knelli2 — CHANGES_REQUESTED** [review]
Just a couple things

File: src/ParallelAlgorithms/Actions/FunctionsOfTimeAreReady.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Actions/FunctionsOfTimeAreReady.hpp]: Yep, I'll also update the simple action variant to say forwarded as well!

File: src/Parallel/Callback.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Parallel/Callback.hpp:103]: Not a problem! Done!

**Review by knelli2 — APPROVED** [review]

**Pull request merged**